File: problems/Maximum Product Subarray/maxProduct.cs
public class Solution {
    public int MaxProduct(int[] nums) {
        int n = nums.Length;
        int num = 1;
        int max = int.MinValue;
        int[] maxArr = new int[n];
        int[] minArr = new int[n];

        for (int i = 0; n > i; ++i) {
            maxArr[i] = nums[i];
            minArr[i] = nums[i];

            if (0 < i) {
                maxArr[i] = Math.Max(maxArr[i], nums[i] * maxArr[i - 1]);
                maxArr[i] = Math.Max(maxArr[i], nums[i] * minArr[i - 1]);

                minArr[i] = Math.Min(minArr[i], nums[i] * maxArr[i - 1]);
                minArr[i] = Math.Min(minArr[i], nums[i] * minArr[i - 1]);
            }

            if (maxArr[i] > max) {
                max = maxArr[i];
            }
        }

        return max;
    }
}

File: problems/Maximum Product of Splitted Binary Tree/maxProduct.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxProduct(TreeNode root) {
        long sum = getSum(root);

        dfs(root, sum);

        return (int)(result%1_000_000_007);
    }

    long result = 0;
    HashMap<TreeNode, Long> nodeToSum = new HashMap<>();

    void dfs(TreeNode root, long totalSum) {
        if (null == root) {
            return;
        }

        long subTreeSum = nodeToSum.get(root);
        long product = subTreeSum * (totalSum - subTreeSum);

        result = Math.max(result, product);

        dfs(root.left, totalSum);
        dfs(root.right, totalSum);
    }

    long getSum(TreeNode root) {
        if (null == root) {
            return 0;
        }

        long leftSum = getSum(root.left);
        long rightSum = getSum(root.right);

        nodeToSum.put(root, leftSum + root.val + rightSum);

        return leftSum + root.val + rightSum;
    }
}
